lib/lexer/lexer.cpp

```diff
@@ -66,64 +77,81 @@ std::function<Lexer::State(Lexer &)> Lexer::getStateHandler() {
 	char ch = chars.peek();
-	if (ch == EOF) {
-		return State::Done;
-	} else if (ch == '\n') {
-		delimit(Token {Token::Type::Newline, Newline()});
-		chars.consume();
-		return State::Empty;
-	} else if (isSpace(ch)) {
-		chars.consume();
-		return State::Empty;
-	} else if (isOperatorStart(ch)) {
-		return State::Operator;
-	} else {
-		return State::Word;
+	while (true) {
+		char ch = chars.peek();
+		if (ch == EOF) {
+			return State::Done;
+		} else if (ch == '\n') {
+			delimit(Token {Token::Type::Newline, Newline()});
+			chars.consume();
+			return State::Empty;
+		} else if (isSpace(ch)) {
+			chars.consume();
+		} else if (isOperatorStart(ch)) {
+			return State::Operator;
+		} else {
+			return State::Word;
+		}
 	}
 }
 
+/*
+Operator -> delimit + switch state
+Backslash -> push state + switch state
+(D|S)Quote/Expansion/Backtick -> pushChar + consume + pushState + switch state
+Space || EOF -> delimit + switch to Empty
+Comment -> delimit + switch to Comment
+*/
 Lexer::State Lexer::wordState() {
 	State state = State::Word;
-	char ch = chars.peek();
-	if (isSpace(ch) || ch == EOF) {
-		state = State::Empty;
-	} else if (isCommentStart(ch)) {
-		state = State::Comment;
-	} else if (isOperatorStart(ch)) {
-		state = State::Operator;
-	} else if (isCommentStart(ch)) {
-		state = State::Comment;
-	} else if (isBackslash(ch)) {
-		state_data.previous = State::Word;
-		return State::Backslash;
-	} else if (isSingleQuote(ch)) {
-		return State::SingleQuoteStart;
-	} else if (isDoubleQuote(ch)) {
-		return State::DoubleQuoteStart;
-	} else if (ch == '$') {
-		return State::ExpansionStart;
-	} else if (ch == '`') {
-		throw NotImplementedException();
-		// state_data.word.push_back(chars.consume());
-		// return State::CommandSubstitution;
-	} else {
-		state_data.word.push_back(ch);
-		chars.consume();
-		return State::Word;
-	}
+	while (state == State::Word) {
+		char ch = chars.peek();
 
+		// [ push char, consume, push state ], switch state
+		if (isSingleQuote(ch) || isDoubleQuote(ch) || isDollarSign(ch) || isBackTick(ch)) {
+			state_data.word.push_back(chars.consume());
+			state_data.pushState(State::Word);
+			if (isSingleQuote(ch)) {
+				return State::SingleQuote;
+			} else if (isDoubleQuote(ch)) {
+				return State::DoubleQuote;
+			} else if (isDollarSign(ch)) {
+				return State::ExpansionStart;
+			} else if (isBackTick(ch)) {
+				return State::BackTick;
+			}
+		}
+		// [ push state, switch state ]
+		if (isBackslash(ch)) {
+			state_data.pushState(State::Word);
+			return State::Backslash;
+		}
+
+		// [ delimit ], switch state
+		if (isSpace(ch) || ch == EOF) {
+			state = State::Empty;
+		} else if (isCommentStart(ch)) {
+			state = State::Comment;
+		} else if (isOperatorStart(ch)) {
+			state = State::Operator;
+		} else {
+			// Continue word state
+			state_data.word.push_back(chars.consume());
+		}
+	}
+	// Handle delimit outside of loop: EOF, IFS, operator, comment
 	if (!state_data.word.empty()) {
```

Comment:
I think I am responsible for this `if statement`, can you check if this is necessary?
The state should never become Word if the character can't become part of a WordToken

Reply:
I believe the only case where it can be empty is a backslash <newline> sequence surrounded by only whitespace, which goes through the `Word` state.

Reply:
So it should be there, since in that case there is no word to delimit, but it's also nice that the empty state doesn't know about this behavior.